File: src/test/java/com/mparticle/kits/AppsflyerKitTests.java
package com.mparticle.kits;


import android.content.Context;

import com.mparticle.MParticle;
import com.mparticle.commerce.CommerceEvent;
import com.mparticle.commerce.Product;
import com.mparticle.commerce.TransactionAttributes;

import org.junit.Test;
import org.mockito.Mockito;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class AppsflyerKitTests {

    private KitIntegration getKit() {
        return new AppsFlyerKit();
    }

    @Test
    public void testGetName() throws Exception {
        String name = getKit().getName();
        assertTrue(name != null && name.length() > 0);
    }

    /**
     * Kit *should* throw an exception when they're initialized with the wrong settings.
     *
     */
    @Test
    public void testOnKitCreate() throws Exception{
        Exception e = null;
        try {
            KitIntegration kit = getKit();
            Map settings = new HashMap<>();
            settings.put("fake setting", "fake");
            kit.onKitCreate(settings, Mockito.mock(Context.class));
        }catch (Exception ex) {
            e = ex;
        }
        assertNotNull(e);
    }

    @Test
    public void testClassName() throws Exception {
        KitIntegrationFactory factory = new KitIntegrationFactory();
        Map<Integer, String> integrations = factory.getKnownIntegrations();
        String className = getKit().getClass().getName();
        for (Map.Entry<Integer, String> entry : integrations.entrySet()) {
            if (entry.getValue().equals(className)) {
                return;
            }
        }
        fail(className + " not found as a known integration.");
    }

    @Test
    public void testGenerateSkuString() throws Exception {
        MParticle.setInstance(Mockito.mock(MParticle.class));
        Mockito.when(MParticle.getInstance().getEnvironment()).thenReturn(MParticle.Environment.Production);
        assertNull(AppsFlyerKit.generateProductIdList(null));
        Product product = new Product.Builder("foo-name", "foo-sku", 50).build();
        CommerceEvent event = new CommerceEvent.Builder(Product.PURCHASE, product)
                .transactionAttributes(new TransactionAttributes("foo"))
                .build();
        assertEquals("foo-sku", AppsFlyerKit.generateProductIdList(event));

        Product product2 = new Product.Builder("foo-name-2", "foo-sku-2", 50).build();
        CommerceEvent event2 = new CommerceEvent.Builder(Product.PURCHASE, product)
                .addProduct(product2)
                .transactionAttributes(new TransactionAttributes("foo"))
                .build();
        assertEquals("foo-sku,foo-sku-2", AppsFlyerKit.generateProductIdList(event2));

        Product product3 = new Product.Builder("foo-name-3", "foo-sku-,3", 50).build();
        CommerceEvent event3 = new CommerceEvent.Builder(Product.PURCHASE, product)
                .addProduct(product2)
                .addProduct(product3)
                .transactionAttributes(new TransactionAttributes("foo"))
                .build();
        assertEquals("foo-sku,foo-sku-2,foo-sku-%2C3", AppsFlyerKit.generateProductIdList(event3));
    }
}

File: src/test/java/com/mparticle/kits/AppsflyerKitTests.kt
package com.mparticle.kits

import android.content.Context
import com.mparticle.MParticle
import com.mparticle.commerce.CommerceEvent
import com.mparticle.commerce.Product
import com.mparticle.commerce.TransactionAttributes
import junit.framework.Assert.assertEquals
import org.junit.Assert
import org.junit.Test
import org.mockito.Mockito


class AppsflyerKitTests {

    private val kit = AppsFlyerKit()

    @Test
    @Throws(Exception::class)
    fun testGetName() {
        val name = kit.name
        Assert.assertTrue(name.isNotEmpty())
    }

    /**
     * Kit *should* throw an exception when they're initialized with the wrong settings.
     *
     */
    @Test
    @Throws(Exception::class)
    fun testOnKitCreate() {
        var e: Throwable? = null
        try {
            val settings = HashMap<String,String>()
            settings["fake setting"] = "fake"
            kit.onKitCreate(settings as Map<String?, String?>, Mockito.mock(Context::class.java))
        } catch (ex: Throwable) {
            e = ex
        }
        Assert.assertNotNull(e)
    }

    @Test
    @Throws(Exception::class)
    fun testClassName() {
        val factory = KitIntegrationFactory()
        val integrations = factory.knownIntegrations
        val className = kit.javaClass.name
        for (integration in integrations) {
            if (integration.value == className) {
                return
            }
        }
        Assert.fail("$className not found as a known integration.")
    }

    @Test
    @Throws(Exception::class)
    fun testGenerateSkuString() {
        MParticle.setInstance(Mockito.mock(MParticle::class.java))
        Mockito.`when`(MParticle.getInstance()?.environment)
            .thenReturn(MParticle.Environment.Production)
        Assert.assertNull(AppsFlyerKit.generateProductIdList(null))
        val product = Product.Builder("foo-name", "foo-sku", 50.0).build()
        val event = CommerceEvent.Builder(Product.PURCHASE, product)
            .transactionAttributes(TransactionAttributes("foo"))
            .build()
        assertEquals("foo-sku", AppsFlyerKit.generateProductIdList(event))
        val product2 = Product.Builder("foo-name-2", "foo-sku-2", 50.0).build()
        val event2 = CommerceEvent.Builder(Product.PURCHASE, product)
            .addProduct(product2)
            .transactionAttributes(TransactionAttributes("foo"))
            .build()
        assertEquals("foo-sku,foo-sku-2", AppsFlyerKit.generateProductIdList(event2))
        val product3 = Product.Builder("foo-name-3", "foo-sku-,3", 50.0).build()
        val event3 = CommerceEvent.Builder(Product.PURCHASE, product)
            .addProduct(product2)
            .addProduct(product3)
            .transactionAttributes(TransactionAttributes("foo"))
            .build()
        assertEquals("foo-sku,foo-sku-2,foo-sku-%2C3", AppsFlyerKit.generateProductIdList(event3))
    }
}